### ISSUE TITLE
Food editor links to old value

### ISSUE THREAD

**Issue by h-h-h-h** [davidhealey/waistline#461]
To reproduce:

- Create food: 100 g, 437 kcal. Save.
- Edit food and stay in this mode.
- Linking is turned on. Change food to 12.5 g and refresh (by tapping on left labels). Calories is now ~55 kcal.
- Turn linking off.
- Change food to 1 piece(s) and turn linking on (this also theoretically refreshes). Food's still at ~55 kcal.
- Change food to 2 piece(s) and refresh. Food's now incorrectly at ~9 kcal (that is 12.5 g or 100 g converted to 2 g). It should be at ~109 kcal (which it'll be, if you save before this step).

**Comment by davidhealey** [comment]
> Change food to 1 piece(s) and turn linking on (this also theoretically refreshes).

This is not the current behaviour. Turning on the link button doesn't update the values so any future changes will operate still on the old values.

I don't know if @EmilJunker's recent pull request will also solve this problem...

**Comment by EmilJunker** [comment]
> I don't know if @EmilJunker's recent pull request will also solve this problem...

No, they don't. And I don't think they should.

It would be very unintuitive if the old values (which are used for the internal calculations if the link button is active) are updated when the user toggles the link button on. There are many scenarios where this would have unintended consequences.

**Comment by davidhealey** [comment]
> > I don't know if @EmilJunker's recent pull request will also solve this problem...
> It would be very unintuitive if the old values (which are used for the internal calculations if the link button is active) are updated when the user toggles the link button on. There are many scenarios where this would have unintended consequences.

Try the steps h-h-h-h outlined and see how you feel about it.

**Comment by EmilJunker** [comment]
> Try the steps h-h-h-h outlined and see how you feel about it.

In my opinion, if you change the item unit, you can't expect the link feature to work properly until you save the item and re-open it. The current behavior is allright.

**Comment by davidhealey** [comment]
> > Try the steps h-h-h-h outlined and see how you feel about it.
> 
> In my opinion, if you change the item unit, you can't expect the link feature to work properly until you save the item and re-open it. The current behavior is allright.

He said to change the serving size though, not the unit

**Comment by EmilJunker** [comment]
> He said to change the serving size though, not the unit

Yes, the original item has a quantity of 100 (the unit is `g`) and a calorie value of 437.

If you change the quantity to 2 (the unit now is `pieces`), the calorie value is updated to ~9. This is the expected behavior.

How is it supposed to know how the handle the unit change until you save the item?

**Comment by davidhealey** [comment]
> How is it supposed to know how the handle the unit change until you save the item?

Oh you're right. I didn't realise the unit had changed to pieces. Follow the same steps though keeping the unit as g and it also has an unintuitive behaviour I think.

**Comment by EmilJunker** [comment]
> Follow the same steps though keeping the unit as g and it also has an unintuitive behaviour

You're right, but changing the behavior could easily lead to issues with rounding. For example, consider this:

- Create an item with a serving size of `100` and a salt value of `0.5`.
- Save the item and re-open it for editing.
- Change the serving size to `25` -> the salt value now is `0.13` (due to rounding).
- Turn the link button off and back on. Intuitively, this should change nothing. But suppose this triggers the old values to be updated.
- If you now change the serving size back to `100`, the salt value would be `0.52`.

**Comment by davidhealey** [comment]
> > Follow the same steps though keeping the unit as g and it also has an unintuitive behaviour
> 
> You're right, but changing the behavior could easily lead to issues with rounding. For example, consider this:
> 
>     * Create an item with a serving size of `100` and a salt value of `0.5`.
> 
>     * Save the item and re-open it for editing.
> 
>     * Change the serving size to `25` -> the salt value now is `0.13` (due to rounding).
> 
>     * Turn the link button off and back on. Intuitively, this should change nothing. But suppose this triggers the old values to be updated.
> 
>     * If you now change the serving size back to `100`, the salt value would be `0.52`.

You're right. I seem to remember fighting this kind of issue when I first implemented the link button. I think for now this case can be worked around by users being sensible and accepting the limitations of the system. If anyone has a better idea then they can implement it :)

**Comment by h-h-h-h** [comment]
> > Change food to 1 piece(s) and turn linking on (this also theoretically refreshes).
> 
> This is not the current behaviour. Turning on the link button doesn't update the values so any future changes will operate still on the old values. [@davidhealey]

I meant this should do the same as tapping on the left labels, in conjunction with hiding the keyboard, probably before the link button action is performed. I also could've written:

- Change food to 1 piece(s), tap on left labels to do any theoretical internal refreshing and prove it isn't changing anything in the UI (ensure a consistent state), and then turn linking on.

> > I don't know if @EmilJunker's recent pull request will also solve this problem...
> 
> No, they don't. And I don't think they should.
> 
> It would be very unintuitive if the old values (which are used for the internal calculations if the link button is active) are updated when the user toggles the link button on. There are many scenarios where this would have unintended consequences. [@EmilJunker]

This whole conversation is based on the misconception I clarified above. The bug I describe lies only in the very last step.

> In my opinion, if you change the item unit, you can't expect the link feature to work properly until you save the item and re-open it. The current behavior is allright. [@EmilJunker]

*This* is very unintuitive. Why can't you expect the link feature to work properly at all times? And changing the unit doesn't matter at all, as far as I understand, because the app doesn't understand anything about the unit and it's only about the serving size numbers and turning linking off an on at the right times.

> Yes, the original item has a quantity of 100 (the unit is g) and a calorie value of 437.
> 
> If you change the quantity to 2 (the unit now is pieces), the calorie value is updated to ~9. This is the expected behavior.
> 
> How is it supposed to know how to handle the unit change until you save the item? [@EmilJunker]

The unit change isn't the relevant part here. The relevant part is that you intermediately turned off linking to be able to change the serving size without the nutrition values following. You set a new serving size value that's associated with the nutrition values as they are. And from that point on, no jumping back to a previous association of the nutrition values with an outdated serving size base should happen.

The unit is just text for the user that can be *expected* to be set to a new value in the context of the user setting a new serving size base. But this isn't relevant for the app's code.

> issues with rounding [@EmilJunker]

That's why I wrote [here](https://github.com/davidhealey/waistline/issues/455) that values should *always* be saved *accurately* (not rounded in any way) and just be *displayed* with rounding. This would mean that as long as you don't manually change a value, a value that's the result of a linked calculation (like 0.0005, displayed as 0.001) will go back to it's previous value with sufficient accuracy (the accuracy of the data type) when you revert the change.

**Comment by EmilJunker** [comment]
> > In my opinion, if you change the item unit, you can't expect the link feature to work properly until you save the item and re-open it. The current behavior is allright. [@EmilJunker]
> 
> This is very unintuitive. Why can't you expect the link feature to work properly at all times?

Because of the aforementioned issues this would cause, first and foremost the rounding errors.

> > How is it supposed to know how to handle the unit change until you save the item? [@EmilJunker]
> 
> The unit change isn't the relevant part here

It is true that the behavior is a bit unintuitive even without the unit change, but without the unit change the scenario makes no sense. For example, if an item has 200 calories per 100 g, why would you want to change it to, say... 100 calories per 30 g? This makes no sense. It would only make sense if it was maybe 100 calories per 30 **pieces**, i.e. you also change the unit. 

> You set a new serving size value that's associated with the nutrition values as they are. And from that point on, no jumping back to a previous association of the nutrition values with an outdated serving size base should happen.

But the serving size and values are not outdated. Not until you actually save the changes you made.

> 

> > issues with rounding [@EmilJunker]
> 
> That's why I wrote here that values should always be saved accurately (not rounded in any way) and just be displayed with rounding

This is simply impossible. You cannot have a numeric input field display one value but internally have a different value. Somehow working around this would be extremely tedious. And absolutely not worth the effort for something like this that can be so easily circumvented by simply saving the item and re-opening it.

**Comment by h-h-h-h** [comment]
> You cannot have a numeric input field display one value but internally have a different value.

You can. There's the app backend, and when a short-lived instance of a food editor page is created, its textboxes are populated with backend data. The app would (as it sounds to be implemeneted) need an additional layer of accurate numbers while the food editor is displayed. Why not just use the textboxes [`dataset`s](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset) to store them for the lifetime of the food editor page?

> It is true that the behavior is a bit unintuitive even without the unit change, but without the unit change the scenario makes no sense. For example, if an item has 200 calories per 100 g, why would you want to change it to, say... 100 calories per 30 g? This makes no sense. It would only make sense if it was maybe 100 calories per 30 pieces, i.e. you also change the unit.

This is exactly what I wrote: "The unit is just text for the user that can be *expected* to be set to a new value in the context of the user setting a new serving size base." (And for this, linking needs to be turned off.)

> But the serving size and values are not outdated. Not until you actually save the changes you made.

When I change 12.5 g to 1 piece(s) while linking is turned off, why should further changes to the serving size number, when linking is turned on again, behave as if the unit was still "g"? Because that's exactly what happens. As you confirmed, it doesn't make sense to have a different serving size base with the *same* unit. When the serving size base/serving size number was changed while linking is turned off, the spirit of the code can *expect* that a unit change happened (but it's not *factually* relevant for the code).

And how should a user know or be able to accept it as sensible that they should save the item after changing to 1 piece(s) and directly come back to make another change to 2 piece(s)?

Perhaps the "2 piece(s)" bit is confusing to you. This, of course, would not be a real use case; this is just for simplification (I didn't expect this to be such an issue). A real use case is creating the food with values referring to 100 ml as written on the package; writing down its gross weight in the notes (as weighed by yourself); then, when all is consumed, change 100 ml to net package volume, say, 1000 ml (linked); change 1000 ml to net package weight as calculated with before-after-values, say, 1500 g (unlinked); and change 1500 g to 100 g (linked) to have neat, customary values in the overview.

**Comment by EmilJunker** [comment]
We're going in circles. I'm not saying it can't be done. I'm just pointing out that the changes you are proposing would potentially cause issues, and it would be very tedious to work around them and come up with a good and robust implementation.

> A real use case is creating the food with values referring to 100 ml as written on the package; writing down ... 

No offense, but this seems to be another one of your overly specific use cases.

In my opinion, David summed it up perfectly:

>  I think for now this case can be worked around by users being sensible and accepting the limitations of the system. If anyone has a better idea then they can implement it :)

**Comment by h-h-h-h** [comment]
In this case, I propose to make things more transparent; because users can only be sensible of the limitations of the system if they notice and understand them:

Whenever the serving size (referring exclusively to the number) was changed while linking was turned off, do the following when linking is turned on again: Show an informational toast with text like: "Values are linked to original serving size until you save."

**Comment by EmilJunker** [comment]
> I propose to make things more transparent

Yeah, I agree with that. Displaying such a toast message every time could be a bit annoying, though. I think it would be better to create a wiki page with an FAQ and detailed explanations. The link button is by far not the only feature that needs some explaining, see discussion at #370.

**Comment by h-h-h-h** [comment]
I understand that repeated toasts for things you long understood can be annoying. Do you think, this one would really be a frequent occurrence, though?

Anyway, context-sensitive help is generally the best, I think. I've barely ever consulted a help file for desktop software. And some things just need to be brought to the user's attention. What do you (not limited to this issue) think of displaying context-based help only once (twice? thrice?) (with an ending like "... See online help."), thus needing to maintain a list of flags what context help was already shown?

**Comment by davidhealey** [comment]
In the long term I plan to overhaul the food editor interface quite significantly to make it more user friendly and appealing to less techny people. When I do that I will try and implement a proper solution to this problem. In the meantime unless we get a lot of feedback about it I'm quite happy to leave it as is.

**Comment by h-h-h-h** [comment]
> When I do that I will try and implement a proper solution to this problem. [@davidhealey]

What do you think about having the 100-g/100-ml values that are normally (always?) on the packaging to be more central and persistent in the backend and in the UI? This means, the reference value like 100 g would be on top of the nutrition section in the food editor. (Other reference values like 250 g, if written on the package, could also be used in the app, since they provide more accuracy.) This would also make it easier to check for up-to-date values, e.g., [with TTS](https://github.com/davidhealey/waistline/issues/457), or just by comparing the kcal value with the current one on the packaging.

*Additionally*, there's still serving size, but as a *list* of serving sizes: You could determine that the whole package has X units instead of 100 (unit should probably be locked and derived from the one in the nutrition section); you could add a serving size of 1 piece(s) and determine that this is the equavalent of, e.g., 12.5 units. (Ideally, there would also be a means of converting volume to weight by inputting before- and after-weights, at least in some calculatory way.)

Then, the nutrition section could have a drop-down box at the top where you can choose between the mutable textbox display for, say, 100 g, and immutable display with labels instead of textboxes for the converted values like "1 package" or "1 piece(s)".

When adding a food to the diary, there'd be an adjustable default unit that's used, but you can also choose between, e.g., piece(s)/g/package.

**Comment by EmilJunker** [comment]
@h-h-h-h I'm sorry, but this is yet another example of a feature request for an overly specific use case. Also, implementing this would probably require structural changes in the database which would be a lot of work.

Nothing prevents you from keeping the serving sizes of your food items set to 100g/100ml. And you can use the notes field to write down the package weight and/or typical serving sizes. In fact, this is exactly what I do.

**Comment by h-h-h-h** [comment]
> And you can use the notes field to write down the package weight and/or typical serving sizes. In fact, this is exactly what I do.

Interesting workaround.

**Comment by EmilJunker** [comment]
> Interesting workaround.

It's not a workaround. That's precisely the purpose of the generic notes field. It exists so different people can use it for different things.
